### ISSUE TITLE
Add custom words to spellchecker? 

### ISSUE THREAD

**Issue by hsarji** [holzschu/Carnets#153]
I like the spellchecker nbextension. Is there a way to add custom words? I'm not sure if I can access the files identified at <https://jupyter-contrib-nbextensions.readthedocs.io/en/latest/nbextensions/spellchecker/README.html> in a way that is similar to the one you identified in [issue 152](https://github.com/holzschu/Carnets/issues/152) or which file to change or add.

**Comment by hsarji** [comment]
I found a solution, but can it be automated? 

I used the code from: https://github.com/ipython-contrib/jupyter_contrib_nbextensions/issues/1428

Each time I need to add a word to my custom dictionary, I have to follow the following steps: 

(Step 1) Run the following code 

```
import os.path
from jupyter_core.paths import jupyter_data_dir
from notebook.services.config import ConfigManager

def AddToDictionary(word):

    local_base_url = os.path.join(
        jupyter_data_dir(),
        'nbextensions', 'spellchecker', 'typo', 'dictionaries')

    lang_code = 'en_US'
    personal_dic = 'personal.dic'
    mod_dic = lang_code + '_mod' + '.dic'

    if not os.path.exists(local_base_url):
        print('creating directory {!r}'.format(local_base_url))
        os.makedirs(os.path.realpath(local_base_url))

    #Add the word to the personal dictionary
    with open(local_base_url + '/' + personal_dic, 'a') as file:
        file.write(word + '\n')

    #Combine the personal and language dictionary
    with open(local_base_url + '/' + lang_code + ".dic", 'r') as file:
        old_lines = file.read()
    with open(local_base_url + '/' + personal_dic, 'r') as file:
        new_lines = file.read()
    with open(local_base_url + '/' + mod_dic, 'w') as file:
        file.write(old_lines)
        file.write(new_lines)

    #Make sure the spellchecker is using the modified dictionary
    cm = ConfigManager()
    rel_path = './typo/dictionaries/' + mod_dic
    cm.update('notebook', {'spellchecker': {'dic_url': rel_path}})

```

(Step 2) Run the following function: 

```
AddToDictionary("NewWord")
```

(Step 3) Restart Carnets 

--- 

I tried placing `personal.dic` in `~/Documents/` and modifying it, but that doesn't do the trick. 

Is there a way define the `AddToDictionary` function when Carnets initializes? Is there another way to add custom words?

**Comment by holzschu** [comment]
Hi,
from what I can tell with a quick look at the documentation of spellchecker, 
- you can edit the configuration file to change the dictionary at `~/Library/Jupyter/nbextensions/spellchecker/main.js`, using `!cp ~/Library/Jupyter/nbextensions/spellchecker/main.js ~/Documents/main.js`, editing, and then `!cp  ~/Documents/main.js ~/Library/Jupyter/nbextensions/spellchecker/main.js` to move it back.
- dictionaries are stored in `~/Library/Jupyter/nbextensions/spellchecker/typo/`.

**Comment by holzschu** [comment]
After a bit of experiments, if you change something in the spellchecker configuration using the nbextension configurator, then it gets written to `~/Documents/.jupyter/nbconfig/notebook.json`, and you can edit this file instead (same method: `!cp ~/Documents/.jupyter/nbconfig/notebook.json ~/Documents/notebook.json`, then edit and move back in place).

**Comment by holzschu** [comment]
And, if I understand things correctly, the function `AddToDictionary` changes the dictionary stored locally, so the changes should be permanent *if* the configuration file does not say to download a new dictionary each time the server restarts. So, something like:
- download favorite dictionary to disk
- move dictionary to `~/Library/Jupyter/nbextensions/spellchecker/typo/`
- edit configuration (using nbextension configurator) to use downloaded dictionary instead of on-line dictionary
- add words to downloaded dictionary 
should work.

**Comment by hsarji** [comment]
Thank you! It works perfectly. I can now add words simply editing the `.dic` file in `~/Documents/notebook.json` and overwriting the one under `~/Library/Jupyter/nbextensions/spellchecker/typo/` without having to run the `AddToDictionary` function. 

Because I ran the function before, my dictionary is now `en_US_mod.dic`. I confirmed this in the `notebook.json` file. (Knowing that this file controls the location of the dictionary is very helpful! Thank you.) 

After I add words to `en_US_mod.dic`, I simply run: 

```
!cp ~/Documents/en_US_mod.dic ~/Library/Jupyter/nbextensions/spellchecker/typo/dictionaries/en_US_mod.dic 
```

I exit out of the notebook (or restart Carnets) and my new words are no longer flagged as spelling errors.
